File: android-tv-client/app/src/main/java/com/momo/ktv/tv/data/ApiClient.kt
package com.momo.ktv.tv.data

import android.util.Log
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody

import java.util.concurrent.TimeUnit

class ApiClient(baseURL: String) {
    companion object {
        private const val TAG = "ApiClient"
        const val CLOUD115_UA = "Mozilla/5.0 115Browser/23.9.3.2"

        /** 规范化服务器地址：确保有 http:// 前缀，去除末尾斜杠 */
        fun normalizeURL(url: String): String {
            var result = url.trim()
            if (result.isEmpty()) return result
            if (!result.startsWith("http://") && !result.startsWith("https://")) {
                result = "http://$result"
            }
            return result.trimEnd('/')
        }
    }

    private var baseURL: String = normalizeURL(baseURL)
    private val gson = Gson()
    private val httpClient = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .build()

    fun updateBaseURL(url: String) {
        baseURL = if (url.startsWith("http")) url else "http://$url"
    }

    fun getServerAddress(): String = baseURL.replace("http://", "").replace("https://", "")

    private fun apiURL(path: String): String = "$baseURL$path"

    // ==================== 歌曲 ====================
    suspend fun fetchSongs(query: String = "", artist: String = "", page: Int = 1, pageSize: Int = 50): Pair<List<Song>, Int> =
        withContext(Dispatchers.IO) {
            val params = mutableListOf<String>()
            if (query.isNotEmpty()) params.add("q=${java.net.URLEncoder.encode(query, "UTF-8")}")
            if (artist.isNotEmpty()) params.add("artist=${java.net.URLEncoder.encode(artist, "UTF-8")}")
            params.add("page=$page")
            params.add("pageSize=$pageSize")
            val path = "/api/songs?" + params.joinToString("&")
            try {
                val req = Request.Builder().url(apiURL(path)).get().build()
                val resp = httpClient.newCall(req).execute()
                val body = resp.body?.string() ?: return@withContext Pair(emptyList(), 0)
                // 尝试 {items, total} 格式
                try {
                    val sr = gson.fromJson(body, SearchResponse::class.java)
                    if (sr.items != null) {
                        return@withContext Pair(sr.items, sr.total ?: sr.items.size)
                    }
                } catch (_: Exception) {}
                // 兼容数组格式
                val type = object : TypeToken<List<Song>>() {}.type
                val songs = gson.fromJson<List<Song>>(body, type) ?: emptyList()
                Pair(songs, songs.size)
            } catch (e: Exception) {
                Log.e(TAG, "fetchSongs error: ${e.message}")
                Pair(emptyList(), 0)
            }
        }

    suspend fun fetchArtists(): List<Artist> = withContext(Dispatchers.IO) {
        try {
            val req = Request.Builder().url(apiURL("/api/artists")).get().build()
            val resp = httpClient.newCall(req).execute()
            val body = resp.body?.string() ?: return@withContext emptyList()
            val type = object : TypeToken<List<Artist>>() {}.type
            gson.fromJson(body, type) ?: emptyList()
        } catch (e: Exception) {
            Log.e(TAG, "fetchArtists error: ${e.message}")
            emptyList()
        }
    }

    // ==================== 队列 ====================
    suspend fun fetchQueue(): List<QueueItem> = withContext(Dispatchers.IO) {
        try {
            val req = Request.Builder().url(apiURL("/api/queue")).get().build()
            val resp = httpClient.newCall(req).execute()
            val body = resp.body?.string() ?: return@withContext emptyList()
            val type = object : TypeToken<List<QueueItem>>() {}.type
            gson.fromJson(body, type) ?: emptyList()
        } catch (e: Exception) {
            Log.e(TAG, "fetchQueue error: ${e.message}")
            emptyList()
        }
    }

    suspend fun addToQueue(songId: Int, nickname: String = "TV用户"): Boolean = withContext(Dispatchers.IO) {
        try {
            val json = gson.toJson(mapOf("song_id" to songId, "nickname" to nickname))
            val req = Request.Builder()
                .url(apiURL("/api/queue"))
                .post(RequestBody.create(MediaType.parse("application/json"), json))
                .build()
            val resp = httpClient.newCall(req).execute()
            resp.isSuccessful
        } catch (e: Exception) {
            Log.e(TAG, "addToQueue error: ${e.message}")
            false
        }
    }

    suspend fun removeFromQueue(queueId: Int): Boolean = withContext(Dispatchers.IO) {
        try {
            val req = Request.Builder().url(apiURL("/api/queue/$queueId")).delete().build()
            val resp = httpClient.newCall(req).execute()
            resp.isSuccessful
        } catch (e: Exception) {
            false
        }
    }

    suspend fun topQueue(queueId: Int): Boolean = withContext(Dispatchers.IO) {
        try {
            val req = Request.Builder()
                .url(apiURL("/api/queue/$queueId/top"))
                .post(RequestBody.create(null, ""))
                .build()
            val resp = httpClient.newCall(req).execute()
            resp.isSuccessful
        } catch (e: Exception) {
            false
        }
    }

    // ==================== 播放信息 ====================
    suspend fun fetchSepInfo(songId: Int): SepInfo? = withContext(Dispatchers.IO) {
        try {
            val req = Request.Builder().url(apiURL("/api/songs/$songId/sep-info")).get().build()
            val resp = httpClient.newCall(req).execute()
            val body = resp.body?.string() ?: return@withContext null
            gson.fromJson(body, SepInfo::class.java)
        } catch (e: Exception) {
            Log.e(TAG, "fetchSepInfo error: ${e.message}")
            null
        }
    }

    fun directStreamURL(filepath: String): String {
        val encoded = java.net.URLEncoder.encode(filepath, "UTF-8").replace("+", "%20")
        return apiURL("/api/direct-stream/$encoded")
    }

    fun coverURL(filename: String?): String? {
        if (filename.isNullOrEmpty()) return null
        return apiURL("/cover/$filename")
    }

    fun wsURL(): String {
        if (baseURL.startsWith("https://")) return baseURL.replace("https://", "wss://") + "/ws"
        return baseURL.replace("http://", "ws://") + "/ws"
    }

    // ==================== 统计 ====================
    suspend fun fetchStats(): Stats? = withContext(Dispatchers.IO) {
        try {
            val req = Request.Builder().url(apiURL("/api/stats")).get().build()
            val resp = httpClient.newCall(req).execute()
            val body = resp.body?.string() ?: return@withContext null
            gson.fromJson(body, Stats::class.java)
        } catch (e: Exception) {
            null
        }
    }
}

File: ios-client/MomoKtviOS/APIClient.swift
import Foundation

/// REST API 客户端：歌曲列表、队列、sep-info、direct-stream 等
class KTVAPIClient {
    static let shared = KTVAPIClient(baseURL: "")

    private(set) var baseURL: String
    let clientId = UUID().uuidString
    let deviceId: String = {
        let key = "momo_ktv_device_id"
        if let existing = UserDefaults.standard.string(forKey: key), !existing.isEmpty {
            return existing
        }
        let newId = "ios-\(UUID().uuidString.prefix(8))"
        UserDefaults.standard.set(newId, forKey: key)
        return newId
    }()

    /// 115 网盘专用 UA
    static let cloud115UserAgent = "Mozilla/5.0 115Browser/23.9.3.2"

    init(baseURL: String) {
        self.baseURL = baseURL.hasPrefix("http") ? baseURL : "http://\(baseURL)"
    }

    func updateBaseURL(_ url: String) {
        self.baseURL = url.hasPrefix("http") ? url : "http://\(url)"
    }

    var serverAddress: String {
        baseURL.replacingOccurrences(of: "http://", with: "")
            .replacingOccurrences(of: "https://", with: "")
    }

    func apiURL(_ path: String) -> URL? {
        URL(string: "\(baseURL)\(path)")
    }

    // MARK: - 歌曲
    func fetchSongs(query: String = "", page: Int = 1, pageSize: Int = 50,
                    completion: @escaping ([Song], Int) -> Void) {
        var params: [String] = ["page=\(page)", "pageSize=\(pageSize)"]
        if !query.isEmpty {
            params.append("q=\(query.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? "")")
        }
        let path = "/api/songs?" + params.joined(separator: "&")
        guard let url = apiURL(path) else { completion([], 0); return }
        URLSession.shared.dataTask(with: url) { data, _, _ in
            guard let data = data else { completion([], 0); return }
            DispatchQueue.main.async {
                if let sr = try? JSONDecoder().decode(SearchResponse.self, from: data),
                   let items = sr.items {
                    completion(items, sr.total ?? items.count)
                } else if let songs = try? JSONDecoder().decode([Song].self, from: data) {
                    completion(songs, songs.count)
                } else {
                    completion([], 0)
                }
            }
        }.resume()
    }

    func fetchArtists(completion: @escaping ([Artist]) -> Void) {
        guard let url = apiURL("/api/artists") else { completion([]); return }
        URLSession.shared.dataTask(with: url) { data, _, _ in
            guard let data = data else { completion([]); return }
            DispatchQueue.main.async {
                completion((try? JSONDecoder().decode([Artist].self, from: data)) ?? [])
            }
        }.resume()
    }

    // MARK: - 队列
    func fetchQueue(completion: @escaping ([QueueItem]) -> Void) {
        guard let url = apiURL("/api/queue") else { completion([]); return }
        URLSession.shared.dataTask(with: url) { data, _, _ in
            guard let data = data else { completion([]); return }
            DispatchQueue.main.async {
                completion((try? JSONDecoder().decode([QueueItem].self, from: data)) ?? [])
            }
        }.resume()
    }

    func addToQueue(songId: Int, nickname: String = "iOS用户", completion: ((Bool) -> Void)? = nil) {
        guard let url = apiURL("/api/queue") else { completion?(false); return }
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try? JSONSerialization.data(withJSONObject: ["song_id": songId, "nickname": nickname])
        URLSession.shared.dataTask(with: req) { _, resp, _ in
            DispatchQueue.main.async {
                completion?((resp as? HTTPURLResponse)?.statusCode == 200)
            }
        }.resume()
    }

    func removeFromQueue(queueId: Int, completion: ((Bool) -> Void)? = nil) {
        guard let url = apiURL("/api/queue/\(queueId)") else { completion?(false); return }
        var req = URLRequest(url: url)
        req.httpMethod = "DELETE"
        URLSession.shared.dataTask(with: req) { _, resp, _ in
            DispatchQueue.main.async { completion?((resp as? HTTPURLResponse)?.statusCode == 200) }
        }.resume()
    }

    func topQueue(queueId: Int, completion: ((Bool) -> Void)? = nil) {
        guard let url = apiURL("/api/queue/\(queueId)/top") else { completion?(false); return }
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        URLSession.shared.dataTask(with: req) { _, resp, _ in
            DispatchQueue.main.async { completion?((resp as? HTTPURLResponse)?.statusCode == 200) }
        }.resume()
    }

    // MARK: - 播放信息
    func fetchSepInfo(songId: Int, completion: @escaping (SepInfo?) -> Void) {
        guard let url = apiURL("/api/songs/\(songId)/sep-info") else { completion(nil); return }
        URLSession.shared.dataTask(with: url) { data, _, _ in
            guard let data = data else { completion(nil); return }
            DispatchQueue.main.async {
                completion(try? JSONDecoder().decode(SepInfo.self, from: data))
            }
        }.resume()
    }

    /// direct-stream URL：服务端 302 重定向到 115 CDN 直链
    func directStreamURL(filepath: String) -> URL? {
        let encoded = filepath.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? filepath
        return apiURL("/api/direct-stream/\(encoded)")
    }

    func coverURL(_ filename: String?) -> URL? {
        guard let filename = filename, !filename.isEmpty else { return nil }
        return apiURL("/cover/\(filename)")
    }

    func wsURL() -> URL? {
        URL(string: baseURL.replacingOccurrences(of: "http", with: "ws") + "/ws")
    }

    func fetchStats(completion: @escaping (Stats?) -> Void) {
        guard let url = apiURL("/api/stats") else { completion(nil); return }
        URLSession.shared.dataTask(with: url) { data, _, _ in
            guard let data = data else { completion(nil); return }
            DispatchQueue.main.async { completion(try? JSONDecoder().decode(Stats.self, from: data)) }
        }.resume()
    }
}
